spark/src/main/scala/geotrellis/spark/io/json/KeyIndexFormats.scala

```diff
@@ -0,0 +1,268 @@
+package geotrellis.spark.io.json
+
+import geotrellis.spark._
+import geotrellis.spark.io.index._
+import geotrellis.spark.io.index.hilbert._
+import geotrellis.spark.io.index.rowmajor._
+import geotrellis.spark.io.index.zcurve._
+
+import com.typesafe.config.ConfigFactory
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+
+import scala.collection.mutable
+import scala.reflect._
+
+import KeyFormats._
+
+case class KeyIndexFormatEntry[K: JsonFormat: ClassTag, T <: KeyIndex[K]: JsonFormat: ClassTag](typeName: String) {
+  val keyClassTag = implicitly[ClassTag[K]]
+  val jsonFormat = implicitly[JsonFormat[T]]
+  val indexClassTag = implicitly[ClassTag[T]]
+}
+
+trait KeyIndexRegistrator {
+  def register(keyIndexRegistry: KeyIndexRegistry)
+}
+
+class KeyIndexJsonFormat[K](entries: Seq[KeyIndexFormatEntry[K, _]]) extends RootJsonFormat[KeyIndex[K]] {
+  def write(obj: KeyIndex[K]): JsValue =
+    entries.find(_.indexClassTag.runtimeClass.isAssignableFrom(obj.getClass)) match {
```

Comment:
Maybe too paranoid, but there is a possible failure case where more than one type is assignable from the the value we got. This should either be called out in the comment as a feature or we should do a `.filter` and throw on multiple matches.